org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/CaseStatement.java

```diff
@@ -506,7 +506,7 @@ private Constant resolveConstantExpression(BlockScope scope,
 			// The following code is copied from InstanceOfExpression#resolve()
 			// But there are enough differences to warrant a copy
 			if (!pb.isReifiable()) {
-				if (expressionType != TypeBinding.NULL) {
+				if (expressionType != TypeBinding.NULL && !(e instanceof RecordPattern)) {
```

Comment:
I don't see why these checks should be appropriate for record patterns: casting to a generic record type will always be "unsafe", but this is compensated by separate instanceof check for the component(s).